core/instantiate.ml

```diff
@@ -28,147 +28,173 @@ type instantiation_maps = (datatype IntMap.t * row IntMap.t * field_spec IntMap.
-        | `Absent -> `Absent
-        | `Var point ->
-            begin
-              match Unionfind.find point with
-                | `Var (var, _, _) ->
-                    if IntMap.mem var penv then
-                      IntMap.find var penv
-                    else
-                      `Var point
-                | `Body f ->
-                    inst_presence rec_env f
-            end
-    and inst_row : inst_env -> row -> row = fun rec_env row ->
-      (* BUG?
-
-         If we change this to unwrap_row then it sometimes leads to
-         divergence during type inference. Is this the correct
-         behaviour? Why does this happen? *)
-      let field_env, row_var, dual = flatten_row row in
-
-      let is_closed =
-        match Unionfind.find row_var with
-          | `Closed -> true
-          | _ -> false in
-
-      let field_env' = StringMap.fold
-        (fun label f field_env' ->
-           let rec add =
-             function
-               | `Present t -> StringMap.add label (`Present (inst rec_env t)) field_env'
-               | `Absent ->
-                   if is_closed then field_env'
-                   else StringMap.add label `Absent field_env'
-               | `Var point ->
-                   match Unionfind.find point with
-                     | `Var (var, _, _) ->
-                         let f =
-                           if IntMap.mem var penv then
-                             IntMap.find var penv
-                           else
-                             `Var point
-                         in
-                           StringMap.add label f field_env'
-                     | `Body f ->
-                         add f
+  let rec inst_typ : instantiation_maps -> inst_env -> datatype -> datatype = fun inst_maps rec_env datatype ->
+    let inst = inst_typ inst_maps rec_env in
+    let instr = inst_row inst_maps rec_env in
+    let tenv, _, _ = inst_maps in
+    let rec_type_env, rec_row_env = rec_env in
+      match datatype with
+        | `Not_typed -> raise (internal_error "`Not_typed' passed to `instantiate'")
+        | `Primitive _  -> datatype
+        | `MetaTypeVar point ->
+            let t = Unionfind.find point in
+              begin
+                match t with
+                  | `Var (var, _, _) ->
+                      if IntMap.mem var tenv then
+                        IntMap.find var tenv
+                      else
+                        datatype
+                  | `Recursive (var, t) ->
+                      Debug.if_set (show_recursion) (fun () -> "rec (instantiate)1: " ^(string_of_int var));
+
+                      if IntMap.mem var rec_type_env then
+                        (`MetaTypeVar (IntMap.find var rec_type_env))
+                      else
+                        begin
+                          let var' = Types.fresh_raw_variable () in
+                          let point' = Unionfind.fresh (`Var (var', (lin_any, res_any), `Flexible)) in
```

Comment:
`Types.fresh_row_variable (lin_any, res_any)`

Reply:
Interesting, that doesn't quite work. `Types.fresh_row_variable` creates a `row meta_row_var_basis`, whereas we need a `typ meta_type_var_basis` here. They differ in that the former also has `Closed` as a constructor. And `Types.fresh_type_variable` wraps the resulting `Var` in `MetaTypeVar`, so you get a type, not just `typ meta_type_var_basis`, so that doesn't quite work, either.

I'll just leave it as it was, rather than adding another helper for creating fresh variables... there are already too many of them anyway.

Reply:
Ugh. I should sit down and clean up this mess one day. Leave it as is for now.

Reply:
I imagine that in most cases, that behavior is actually what we want. So it's just the naming that's slightly inconsistent, but not a big deal. And after the big refactoring of types, this distinction between row and type vars will be gone anyways.